Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,184 @@
+Authors: Joe Hendrix, Wojciech Nawrocki, Leonardo de Moura
+-/
+namespace Std
+/-!
+We define bitvectors. We choose the `Fin` representation over others for its relative efficiency
+(Lean has special support for `Nat`), alignment with `UIntXY` types which are also represented
+with `Fin`, and the fact that bitwise operations on `Fin` are already defined. Some other possible
+representations are `List Bool`, `{ l : List Bool // l.length = w }`, `Fin w → Bool`.
+
+We also define many bitvector operations from the
+[`QF_BV` logic](https://smtlib.cs.uiowa.edu/logics-all.shtml#QF_BV).
+of SMT-LIBv2.
+
+TODO(Leo): match the interface to what SMT-LIB provides
+-/
+
+/--
+A bitvector of the specified width. This is represented as the underlying `Nat` number
+in both the runtime and the kernel, inheriting all the special support for `Nat`.
+-/
+structure BitVec (w : Nat) where
+  /-- Bitvector is represented as a `Fin`. -/
+  val : Fin (2^w)
+  deriving DecidableEq
+
+namespace BitVec
+
+theorem pow_two_gt_zero (w : Nat) : 2^w > 0 := by
+  apply Nat.pos_pow_of_pos; decide
+
+/-- The BitVector `i mod 2^n`. -/
+protected def ofNat (n : Nat) (i : Nat) : BitVec n :=
+  { val := Fin.ofNat' i (pow_two_gt_zero _) }
+
+/-- Given a bitvector `a`, return the underlying `Nat`. -/
+protected def toNat (a : BitVec n) : Nat :=
+  a.val.val
+
+/-- Return a bitvector `0` of size `n`.  -/
+protected def zero (n : Nat) : BitVec n :=
+  BitVec.ofNat n 0
+
+instance : Inhabited (BitVec n) where
+  default := BitVec.zero n
+
+instance : OfNat (BitVec n) i where
+  ofNat := BitVec.ofNat n i
+
+/--
+Notation for bit vector literals. `i#n` is a shorthand for `BitVec.ofNat n i`.
+TODO: is this a good notation? Do we need it? We can also write `(i : BitVec n)`
+-/
+scoped syntax:max num noWs "#" noWs num : term
+macro_rules | `($i:num#$n:num) => `(BitVec.ofNat $n $i)
+
+/-- Unexpander for bit vector literals. -/
+@[app_unexpander BitVec.ofNat] def unexpandBitVecOfNat : Lean.PrettyPrinter.Unexpander
+  | `($(_) $n:num $i:num) => `($i:num#$n:num)
+  | _                => throw ()
+
+instance : Repr (BitVec n) where
+  reprPrec a _ := repr a.val ++ "#" ++ repr n
+
+instance : ToString (BitVec n) where
+  toString a := toString (repr a)
+
+/-- Theorem for normalizing the bit vector literal representation. -/
+@[simp] theorem ofNat_eq_ofNat : @OfNat.ofNat (BitVec n) i _ = BitVec.ofNat n i := rfl
+
+/-- Addition for bit vectors. -/
+protected def add (x y : BitVec n) : BitVec n :=
+  { val := x.val + y.val }
+/-- Subtraction for bit vectors. -/
+protected def sub (x y : BitVec n) : BitVec n :=
+  { val := x.val - y.val }
+/-- Multiplication for bit vectors. -/
+protected def mul (x y : BitVec n) : BitVec n :=
+  { val := x.val * y.val }
+/-- Modulo for bit vectors. -/
+protected def mod (x y : BitVec n) : BitVec n :=
+  { val := x.val % y.val }
+/-- Division for bit vectors. -/
+protected def div (x y : BitVec n) : BitVec n :=
+  { val := x.val / y.val }
```

Comment:
Should we be compatible with SMT-LIB for division by zero?  Specifically, SMT-LIB defines `x / 0` on bitvectors to be all ones (see the [spec](http://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml)).  This is consistent with a simple division circuit.

Here I am getting `0xABCD#16 / 0#16` to be `0`.

Reply:
Yes, we should. Good catch.

Reply:
We certainly at least want an `smtDiv` in that case. I think we should bias the notation to whatever leads to shorter theorems, but I don't think we have enough data on what that is yet. I have a slight preference for consistency with other lean definitions since then theorem transfer will be easier.

Reply:
@digama0 Thanks for sharing your position. Nothing is final here and all decisions can be reverted later. We will move forward quickly and get feedback for the `BitVec` library from people that will be using Lean in software/hardware verification projects. Having industrial applications and users is super important for the Lean project future. 
The good news is that the work on `BitVec` does not affect Mathlib.

Reply:
@leodemoura I will have you know that I don't need to be reminded how "we will move forward quickly". As a maintainer, I'm feeling pretty bulldozed right now. Please allow me to do due diligence in keeping the high quality standards of this library. I assure you it won't take that long; I just want to make sure that the rationales are all written down (either here or in the code itself) so that we can examine them later if we need to make changes in the future.

Reply:
@digama0 I want to start by expressing my deep appreciation for everything you've done for Std4. Your commitment and efforts since stepping up as its maintainer last year, when I ask you for help, have been nothing short of exemplary. I recognize the weight of responsibilities that came with it, and I genuinely regret if my recent comments added any undue pressure.

Please know that your concerns about maintaining the high quality of this library resonate with me. I've always valued your dedication to thoroughness and meticulousness. The intention behind mentioning moving forward quickly was never to undermine your role or the due diligence you bring.

As I had mentioned [here](https://github.com/leanprover/lean4/pull/2542#issuecomment-1723352703), help was coming. I did not forget the burden you were carrying :) Joe has now taken on the responsibility of being the full-time maintainer for Std4, aiming to progress the project further. This, however, doesn’t diminish the value of your insights or your role. You'll continue to have maintainer privileges, and your feedback will always be invaluable to us.

If you ever feel the need to lighten your workload, please know that the title of 'honorary maintainer' is open to you. This isn't about reducing your role but rather an acknowledgment of the immense work you've done and a gesture of gratitude. The priority here is your well-being and ensuring that you feel respected and acknowledged.

Thank you for everything, Mario.

Reply:
I appreciate that. I would rather not step down from the maintainer role though, I'd rather be an "honorary FRO member" and retain code ownership over my work. I'm glad Joe is coming on board, two maintainers are better than one.

Reply:
I love that. How are your expectations regarding the "Honorary FRO member" role? Happy to sync with you on a call this week.